backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,125 @@
+// Copyright © 2022 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+const DefaultViewRefreshIntervalSecond = 5
+
+type refreshFunc func(db *gorm.DB, viewNames []string)
+
+type ViewRefreshHandler struct {
+	mu                        sync.Mutex
+	viewsToRefresh            map[string][]string
+	tableChanged              map[string]bool
+	refreshFunc               refreshFunc
+	viewRefreshIntervalSecond time.Duration
+}
+
+func (vh *ViewRefreshHandler) TableChanged(table string) {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.tableChanged[table] = true
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	vh.refreshFunc(db, viewNames)
+}
+
+func (vh *ViewRefreshHandler) getViewsToRefresh() []string {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table, ok := range tables {
+		if ok {
+			for _, viewName := range vh.viewsToRefresh[table] {
+				viewToRefresh[viewName] = true
```

Comment:
why there is a need to use `viewToRefresh` map and not creating the `views` slice here?

If this is done in order to remove duplication a comment here will be nice.

Reply:
@FrimIdan the viewToRefresh map here is actually being used as a set, if two tables changed which require the same view then we don't want to post-process a slice to remove the duplicates for the next step

Reply:
Yep, so a comment in the code will be nice IMO

Reply:
i've added comments